File: apps/web/src/lib/documents.ts
import { createClient } from "@/lib/supabase/client";

/** US-2.21/2.22: browser-side document CRUD — straight to Supabase under
 * RLS ("build less API"). Objects live in the private `project-docs`
 * bucket at `<org_id>/projects/<project_id>/<document_id>/<filename>`;
 * metadata in public.documents. Factory/agent writes go through `api`. */

export const DOCS_BUCKET = "project-docs";
export const MAX_DOCUMENT_BYTES = 25 * 1024 * 1024;

export type DocumentSource = "user" | "factory" | "agent";
export type DocumentAttachedTo = "project" | "work-item" | "prd" | "test-case";

export type DocumentRow = {
  id: string;
  org_id: string;
  project_id: string;
  issue_id: string | null;
  test_case_id: string | null;
  run_id: string | null;
  name: string;
  mime_type: string;
  size_bytes: number;
  storage_path: string;
  source: DocumentSource;
  attached_to: DocumentAttachedTo;
  created_by: string | null;
  created_at: string;
  updated_at: string;
};

/** Where an upload lands: exactly one link per document (US-2.22). */
export type DocumentTarget =
  | { attachedTo: "project" }
  | { attachedTo: "work-item" | "prd"; issueId: string }
  | { attachedTo: "test-case"; testCaseId: string };

/** Mirrors the api's safe_name: last path segment, storage-safe chars. */
export function safeName(name: string): string {
  const base = (name || "").replaceAll("\\", "/").split("/").pop()?.trim() ?? "";
  const cleaned = base.replace(/[^A-Za-z0-9._ ()-]/g, "_").replace(/^[. ]+|[. ]+$/g, "");
  return cleaned || "document";
}

export function formatBytes(n: number): string {
  if (n < 1024) return `${n} B`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} KB`;
  return `${(n / (1024 * 1024)).toFixed(1)} MB`;
}

export function isPreviewableImage(mime: string): boolean {
  return mime.startsWith("image/");
}

export function isPreviewableHtml(mime: string): boolean {
  return mime === "text/html";
}

function targetRefs(target: DocumentTarget): {
  issue_id: string | null;
  test_case_id: string | null;
} {
  return {
    issue_id: "issueId" in target ? target.issueId : null,
    test_case_id: "testCaseId" in target ? target.testCaseId : null,
  };
}

export function matchesTarget(doc: DocumentRow, target: DocumentTarget): boolean {
  if (doc.attached_to !== target.attachedTo) return false;
  const refs = targetRefs(target);
  return (
    doc.issue_id === refs.issue_id && doc.test_case_id === refs.test_case_id
  );
}

export async function uploadDocument(
  orgId: string,
  projectId: string,
  target: DocumentTarget,
  file: File
): Promise<DocumentRow> {
  if (file.size > MAX_DOCUMENT_BYTES) {
    throw new Error("Documents are limited to 25 MB.");
  }
  const supabase = createClient();
  const {
    data: { user },
  } = await supabase.auth.getUser();

  const id = crypto.randomUUID();
  const name = safeName(file.name);
  const path = `${orgId}/projects/${projectId}/${id}/${name}`;
  const mime = file.type || "application/octet-stream";

  const { error: uploadError } = await supabase.storage
    .from(DOCS_BUCKET)
    .upload(path, file, { contentType: mime });
  if (uploadError) throw new Error(uploadError.message);

  const { data, error } = await supabase
    .from("documents")
    .insert({
      id,
      org_id: orgId,
      project_id: projectId,
      ...targetRefs(target),
      attached_to: target.attachedTo,
      name,
      mime_type: mime,
      size_bytes: file.size,
      storage_path: path,
      source: "user",
      created_by: user?.id ?? null,
    })
    .select()
    .single();
  if (error) {
    // Don't strand the object behind a failed row insert.
    await supabase.storage.from(DOCS_BUCKET).remove([path]);
    throw new Error(error.message);
  }
  return data as DocumentRow;
}

/** US-2.22 "Replace file": new bytes, same document id and links. The
 * source badge flips to the last writer (here: the user). */
export async function replaceDocumentFile(
  doc: DocumentRow,
  file: File
): Promise<DocumentRow> {
  if (file.size > MAX_DOCUMENT_BYTES) {
    throw new Error("Documents are limited to 25 MB.");
  }
  const supabase = createClient();
  const {
    data: { user },
  } = await supabase.auth.getUser();

  const name = safeName(file.name);
  const mime = file.type || "application/octet-stream";
  const dir = doc.storage_path.slice(0, doc.storage_path.lastIndexOf("/"));
  const path = `${dir}/${name}`;

  const { error: uploadError } = await supabase.storage
    .from(DOCS_BUCKET)
    .upload(path, file, { contentType: mime, upsert: true });
  if (uploadError) throw new Error(uploadError.message);

  if (path !== doc.storage_path) {
    await supabase.storage.from(DOCS_BUCKET).remove([doc.storage_path]);
  }

  const { data, error } = await supabase
    .from("documents")
    .update({
      name,
      mime_type: mime,
      size_bytes: file.size,
      storage_path: path,
      source: "user",
      created_by: user?.id ?? null,
    })
    .eq("id", doc.id)
    .select()
    .single();
  if (error) throw new Error(error.message);
  return data as DocumentRow;
}

/** Delete removes the row and the object. */
export async function deleteDocument(doc: DocumentRow): Promise<void> {
  const supabase = createClient();
  const { error } = await supabase.from("documents").delete().eq("id", doc.id);
  if (error) throw new Error(error.message);
  await supabase.storage.from(DOCS_BUCKET).remove([doc.storage_path]);
}

export async function documentUrl(
  doc: DocumentRow,
  opts?: { download?: boolean }
): Promise<string> {
  const supabase = createClient();
  const { data, error } = await supabase.storage
    .from(DOCS_BUCKET)
    .createSignedUrl(doc.storage_path, 3600, {
      download: opts?.download ? doc.name : undefined,
    });
  if (error || !data) throw new Error(error?.message ?? "Could not sign URL");
  return data.signedUrl;
}

File: apps/api/app/documents.py
"""Project document server-side write/read path (US-2.21/2.22).

Browser CRUD goes straight to Supabase under RLS ("build less API") —
this module exists only for the writers with no user JWT or a non-user
source: factory-generated documents (PRD wireframes), agent uploads
mid-run, and the runner byte-fetch. Objects live in the `project-docs`
bucket at `<org_id>/projects/<project_id>/<document_id>/<filename>`;
metadata in public.documents.

A write with the same name to the same link target replaces the
document in place — same id, links intact — so regenerated wireframes
converge instead of piling up (US-2.22).
"""

from __future__ import annotations

import re
import uuid
from typing import Any

from . import storage
from .config import Settings
from .db import _connect

MAX_DOCUMENT_BYTES = 25 * 1024 * 1024  # mirrors the bucket's file_size_limit

_MIME_BY_EXT = {
    "svg": "image/svg+xml",
    "html": "text/html",
    "htm": "text/html",
    "mmd": "text/plain",
    "md": "text/markdown",
    "txt": "text/plain",
    "json": "application/json",
    "pdf": "application/pdf",
    "png": "image/png",
    "jpg": "image/jpeg",
    "jpeg": "image/jpeg",
    "gif": "image/gif",
    "webp": "image/webp",
}


def safe_name(name: str | None) -> str:
    """A bare filename usable as the last storage-path segment: no
    separators (path traversal) and no characters Storage rejects."""
    base = (name or "").replace("\\", "/").rsplit("/", 1)[-1].strip()
    base = re.sub(r"[^A-Za-z0-9._ ()-]", "_", base).strip(". ")
    return base or "document"


def mime_for(name: str, declared: str | None = None) -> str:
    if declared and declared != "application/octet-stream":
        return declared
    ext = name.rsplit(".", 1)[-1].lower() if "." in name else ""
    return _MIME_BY_EXT.get(ext, "application/octet-stream")


def object_path(org_id: str, project_id: str, document_id: str, name: str) -> str:
    return f"{org_id}/projects/{project_id}/{document_id}/{name}"


_ROW_COLS = (
    "id, org_id, project_id, issue_id, test_case_id, run_id, name, mime_type, "
    "size_bytes, storage_path, source, attached_to, created_by, created_at, updated_at"
)


def _row_dict(row: dict[str, Any]) -> dict[str, Any]:
    return {k: (str(v) if isinstance(v, uuid.UUID) else v) for k, v in row.items()}


def get_document(settings: Settings, document_id: str) -> dict[str, Any] | None:
    try:
        uuid.UUID(document_id)
    except ValueError:
        return None
    with _connect(settings) as conn:
        row = conn.execute(
            f"select {_ROW_COLS} from public.documents where id = %s",
            (document_id,),
        ).fetchone()
    return _row_dict(row) if row else None


def _find_existing(
    conn,
    *,
    project_id: str,
    name: str,
    attached_to: str,
    issue_id: str | None,
    test_case_id: str | None,
) -> dict[str, Any] | None:
    return conn.execute(
        f"""
        select {_ROW_COLS} from public.documents
        where project_id = %s and name = %s and attached_to = %s
          and issue_id is not distinct from %s
          and test_case_id is not distinct from %s
        """,
        (project_id, name, attached_to, issue_id, test_case_id),
    ).fetchone()


async def create_or_replace(
    settings: Settings,
    *,
    org_id: str,
    project_id: str,
    name: str,
    content: bytes,
    source: str,
    attached_to: str = "work-item",
    mime_type: str | None = None,
    issue_id: str | None = None,
    test_case_id: str | None = None,
    run_id: str | None = None,
    created_by: str | None = None,
) -> dict[str, Any]:
    """Stage the object and upsert its metadata row; returns the row."""
    filename = safe_name(name)
    mime = mime_for(filename, mime_type)

    with _connect(settings) as conn:
        existing = _find_existing(
            conn,
            project_id=project_id,
            name=filename,
            attached_to=attached_to,
            issue_id=issue_id,
            test_case_id=test_case_id,
        )
        if existing:
            await storage.put_object(
                settings,
                existing["storage_path"],
                content,
                content_type=mime,
                bucket=storage.DOCS_BUCKET,
            )
            row = conn.execute(
                f"""
                update public.documents
                set mime_type = %s, size_bytes = %s, source = %s,
                    run_id = coalesce(%s, run_id), created_by = %s
                where id = %s
                returning {_ROW_COLS}
                """,
                (mime, len(content), source, run_id, created_by, existing["id"]),
            ).fetchone()
            conn.commit()
            return _row_dict(row)

        doc_id = str(uuid.uuid4())
        path = object_path(org_id, project_id, doc_id, filename)
        await storage.put_object(
            settings, path, content, content_type=mime, bucket=storage.DOCS_BUCKET
        )
        try:
            row = conn.execute(
                f"""
                insert into public.documents
                  (id, org_id, project_id, issue_id, test_case_id, run_id, name,
                   mime_type, size_bytes, storage_path, source, attached_to, created_by)
                values (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                returning {_ROW_COLS}
                """,
                (
                    doc_id,
                    org_id,
                    project_id,
                    issue_id,
                    test_case_id,
                    run_id,
                    filename,
                    mime,
                    len(content),
                    path,
                    source,
                    attached_to,
                    created_by,
                ),
            ).fetchone()
            conn.commit()
        except Exception:
            # Don't strand the staged object behind a failed row insert.
            await storage.delete_object(settings, path, bucket=storage.DOCS_BUCKET)
            raise
        return _row_dict(row)


async def read_bytes(settings: Settings, doc: dict[str, Any]) -> bytes | None:
    return await storage.get_object(
        settings, doc["storage_path"], bucket=storage.DOCS_BUCKET
    )
